### ISSUE TITLE
Can we get consistent defaulting for API?

### ISSUE THREAD

**Issue by thockin** [GoogleCloudPlatform/kubernetes#1502]
In #1458, Brendan introduced a pattern for accessing and defaulting values:  Rather than actually setting the value in the struct during validation (which we do for some other fields), he wrapped the comparisons in functions that captured the defaulting.

This pattern has a distinct advantage of working in tests, where input validation has not run.  It has the distinct disadvantage that not all fields are so self-contained - some set their default based on another field's value.

We should choose a pattern and apply it liberally.  I'd like to be consistent as much as possible.

**Comment by bgrant0607** [comment]
Recap of what I explained to @erictune for @derekwaynecarr wrt discussion in #3472 about whether defaulted resource values belong in PodSpec or in PodStatus:

The desired state (`Spec` in v1beta3) in the system should actually reflect the actual, full desired state. That makes it easier for the system to determine what to do and easier for the user to understand what is going to happen.

Lots of things will set Spec (and metadata) fields of objects in addition to directly user-specified configuration values:
* Config generators
* Config transformations
* Client-side defaulting logic (e.g., using info from images, like podex, or contextual defaulting)
* Version-specific defaulting logic in apiserver (see #2587 for the approach we've settled on)
* Pervasively set metadata, based on kubectl configuration (e.g., for namespace, apiversion), kubectl command-line flags (e.g., `--labels`), etc.
* Imperative CLI updates (e.g., resize, patch, relabel)
* Deployment/rollout tools, like kubectl rollingupdate
* "Decorator" proxies that interpose between clients and the apiserver
* Custom verbs (e.g., resize, stop) and synthetic control objects (e.g., bindings)
* Automatic population of readonly fields, such as creationTimestamp or uid
* Auto-scalers and vertical auto-sizers 
etc. etc.

Therefore, a declarative configuration reconciliation mechanism effectively needs to patch the existing desired state with the fields the user maintains using declarative configuration. The basic approach is described in #1702 and in some of the older issues it references. It proposes keeping track of the fields to merge in annotations.

As another example of using annotations to assist the update process, both Openshift and kubectl use annotations to keep track of the final desired state and/or intermediate state during rolling updates. A method I proposed in order to enable simultaneous resizing by an auto-scaler and rolling updater is described here: https://github.com/GoogleCloudPlatform/kubernetes/pull/2863#issuecomment-69701562.

**Comment by bgrant0607** [comment]
This is mostly done. I did find a stray default:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/registry/service/rest.go#L133

**Comment by bgrant0607** [comment]
We should also document the approach in docs/api-conventions.md.

**Comment by ghodss** [comment]
@bgrant0607 Can you summarize the approach taken for how defaults are handled?

**Comment by thockin** [comment]
@ghodss Defaulting is now done in a separate per-version pass during conversion from versioned APIs to internal structs.

**Comment by bgrant0607** [comment]
@ghodss An example is here:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta1/defaults.go#L27

**Comment by bgrant0607** [comment]
I think this is done.
